c7n/resources/elasticache.py

```diff
@@ -518,6 +517,60 @@ def process_resource_set(self, snapshots, tags):
             client.add_tags_to_resource(ResourceName=arn, Tags=tags)
 
 
+@ElastiCacheSnapshot.action_registry.register('copy-cluster-tags')
+class CopyClusterTags(BaseAction):
+    """
+    Copy specified tags from Elasticache cluster to Snapshot
+    :example:
+
+        .. code-block: yaml
+
+            - name: elasticache-test
+              resource: cache-snapshot
+              filters:
+                 - type: value
+                   key: SnapshotName
+                   op: in
+                   value:
+                    - test-tags-backup
+              actions:
+                - type: copy-cluster-tags
+                  tags:
+                    - tag1
+                    - tag2
+    """
+
+    schema = type_schema(
+        'copy-cluster-tags',
+        tags={'type': 'array', 'items': {'type': 'string'}})
+
+
+    def process(self, snapshots):
+        log.info("Modifying %d ElastiCache snapshots", len(snapshots))
+        cluster_ids = {s['CacheClusterId'] for s in snapshots}
+        clusters = [
+            cluster for cluster in
+            ElastiCacheCluster(self.manager.ctx, {}).resources()
+            if cluster['CacheClusterId'] in cluster_ids]        # get cluster tags
+
+        for s in snapshots:
+            arn = self.manager.generate_arn(s['SnapshotName'])
+            for c in clusters:
```

Comment:
rather than do a nested loop, it would be better to make the cluster list a dictionary on cluster id to cluster and turn this inner loop into a lookup.